File: Algorithms/430. Flatten a Multilevel Doubly Linked List/javascript/solution.recursion.js
/**
 * // Definition for a Node.
 * function Node(val,prev,next,child) {
 *    this.val = val;
 *    this.prev = prev;
 *    this.next = next;
 *    this.child = child;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var flatten = function (head) {
    if (!head) {
        return;
    }

    const flattenDfs = (prev, curr) => {
        if (!curr) {
            return prev;
        }

        prev.next = curr;
        curr.prev = prev;

        let tempNext = curr.next;
        let tail = flattenDfs(curr, curr.child);
        curr.child = null;
        return flattenDfs(tail, tempNext);
    };

    let pseudoHead = new Node(0, null, head, null);
    flattenDfs(pseudoHead, head);
    pseudoHead.next.prev = null;
    return pseudoHead.next;
};

File: Algorithms/430. Flatten a Multilevel Doubly Linked List/golang/solution.recursion.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Prev *Node
 *     Next *Node
 *     Child *Node
 * }
 */

func flatten(root *Node) *Node {
	if root == nil {
		return nil
	}

	var pseudoHead *Node = &Node{
		Next: root,
	}
	flattenDfs(pseudoHead, root)
	pseudoHead.Next.Prev = nil
	return pseudoHead.Next
}

func flattenDfs(prev *Node, curr *Node) *Node {
	if curr == nil {
		return prev
	}

	curr.Prev = prev
	prev.Next = curr

	var tempNext *Node = curr.Next
	var tail *Node = flattenDfs(curr, curr.Child)
	curr.Child = nil
	return flattenDfs(tail, tempNext)
}